metricbeat/module/mongodb/dbstats/dbstats.go

```diff
@@ -0,0 +1,90 @@
+package dbstats
+
+import (
+	"errors"
+
+	"github.com/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/metricbeat/mb"
+	"gopkg.in/mgo.v2"
+)
+
+// init registers the MetricSet with the central registry.
+// The New method will be called after the setup of the module and before starting to fetch data
+func init() {
+	if err := mb.Registry.AddMetricSet("mongodb", "dbstats", New); err != nil {
+		panic(err)
+	}
+}
+
+// MetricSet type defines all fields of the MetricSet
+// As a minimum it must inherit the mb.BaseMetricSet fields, but can be extended with
+// additional entries. These variables can be used to persist data or configuration between
+// multiple fetch calls.
+type MetricSet struct {
+	mb.BaseMetricSet
+	dialInfo *mgo.DialInfo
+}
+
+// New create a new instance of the MetricSet
+// Part of new is also setting up the configuration by processing additional
+// configuration entries if needed.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	dialInfo, err := mgo.ParseURL(base.HostData().URI)
+	if err != nil {
+		return nil, err
+	}
+	dialInfo.Timeout = base.Module().Config().Timeout
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		dialInfo:      dialInfo,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch() ([]common.MapStr, error) {
+
+	// establish connection to mongo
+	session, err := mgo.DialWithInfo(m.dialInfo)
```

Comment:
@ruflin 

## Connecting to single instances

Tricky indeed. Fortunately, mgo (the mongo client) has a `dialInfo.Direct` parameter. In mongobeat, I initiate a master connection to the cluster, and then create individual node connections to each node. You can see it [here](https://github.com/scottcrespo/mongobeat/blob/master/mongo/mongo.go)

I create the individual connections in the "controller" [here](https://github.com/scottcrespo/mongobeat/blob/master/beater/mongobeat.go#L64)

I've begun implementing something similar for the mongodb module, minus node discovery. This is implemented at the module level to share logic and prevent redundant connections.

## Detecting master/primary

Easily implemented with mgo. You can specify read preference to always read from primary. If the primary changes, your reads will automatically be directed to the new primary. 

## dbStats response the same

Technically dbStats response (and serverStatus) are not the same for each node. Sometimes the differences are trivial, other times they're not. I'll cover the cases where they're *not* the same:

**[replica sets](https://docs.mongodb.com/manual/replication/)**

In the case of a replica set, the state of the secondaries lag behind the primary. Thus, seeing the stats for each node helps monitor replication lag and success/failure. This is especially helpful in the case of multi-datacenter replication, where larger discrepancies are usually present.

**[sharded clusters](https://docs.mongodb.com/manual/core/sharded-cluster-shards/)**

In the case of a sharded cluster, the complexity is compounded because nodes contain different databases and collections, and a shard can also be a replica set! =o

## Conclusion

Given these considerations, I'm still having a hard time understanding what the best implementation for metricbeat would be. Given the current implementation - the node that is read from is implicit and can change from period to period. For example, you might connect to mongo on `localhost` and expect that you're reading from that node. But under the hood, mgo is actually discovering the other nodes in the cluster and directing reads to any one of the member nodes. Thus, I think it's possible to receive unexpected and inconsistent results. 

If we enforce that metricbeat only establishes a direct connection to a single mongo instance on localhost, then all of this complexity goes away and output is very predictable. Yay! But, it limits the flexibility of the system. 

If we allow the agent to connect to mongo as a cluster, and permit configuration of multiple hosts, then I think there's a lot of different directions we could take. 

Definitely a Zoom call will be helpful. I'm sure you can teach me a lot more about how metricbeat works under the hood, so I can understand the best direction.  

I don't have your email, but can you contact me at sccrespo@gmail.com and we'll set up a time? Perhaps we could talk sometime Thursday afternoon (your time).